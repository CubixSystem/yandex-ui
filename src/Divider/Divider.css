.Divider {
    display: flex;
    align-items: center;

    box-sizing: border-box;
    width: 100%;
}

.Divider-Line {
    flex: 1 1 0;

    background-color: var(--divider-fill-color);
}

.Divider-Line:first-child:not(:only-child) {
    margin-right: var(--divider-indent-all);
}

.Divider-Line:last-child:not(:only-child) {
    margin-left: var(--divider-indent-all);
}
