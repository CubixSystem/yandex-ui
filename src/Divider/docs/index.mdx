---
id: divider
path: /components/divider

title: Divider
description: Компонент для отделения одной информации от другой.

links:
 - label: source
   url: https://github.yandex-team.ru/search-interfaces/frontend/tree/master/packages/lego-components/src/Divider
 - label: startrek
   url: https://search.yandex-team.ru/stsearch?text=Divider.ts&facet.queue=ISL&facet.type=bug&facet.status=128
 - label: figma
   url: https://www.figma.com/file/CaoHFEO0w6PaE8cVaYFDoq8i/L-E-G-O?node-id=0%3A1

tabs:
 - Examples
 - Usage
 - Playground
 - Props
---
