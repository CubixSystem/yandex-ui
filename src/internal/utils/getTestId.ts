/**
 * Возвращает ID для использования в тестах.
 *
 * @example
 *
 * getTestId('host', testId)
 */
export function getTestId(prefix:string, id?: string): string | undefined {
    if (id === undefined) {
        return undefined;
    }
    return `${id}-${prefix}`;
}
