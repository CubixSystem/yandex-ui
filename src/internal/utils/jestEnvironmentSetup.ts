/**
 * Возвращает DOM-узел для монтирования компонента.
 */
export const getRoot = (): HTMLElement | null => {
    return document.getElementById('root');
};

/**
 * Хелпер, создающий корневой контейнер для монтирования,
 * необходим для того, чтобы избежать ошибок при монтировании в enzyme.
 */
export const jestEnvironmentSetup = (): void => {
    // eslint-disable-next-line no-undef
    beforeEach(() => {
        const node = document.createElement('div');
        node.setAttribute('id', 'root');
        document.body.appendChild(node);
    });

    // eslint-disable-next-line no-undef
    afterEach(() => {
        const node = getRoot();
        if (node !== null) {
            document.body.removeChild(node);
        }
    });
};
