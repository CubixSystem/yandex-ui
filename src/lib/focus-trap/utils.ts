import { getTabbables } from './tabbable';

const FOCUS_TRAP_GUARD_ATTR = 'focusTrapGuard';

interface ScrollableElement {
    element: HTMLElement;
    scrollTop: number;
    scrollLeft: number;
}

let supportsPreventScrollCached: boolean | null = null;
function supportsPreventScroll() {
    if (supportsPreventScrollCached === null) {
        supportsPreventScrollCached = false;

        try {
            const element = document.createElement('div');

            element.focus({
                get preventScroll() {
                    supportsPreventScrollCached = true;
                    return true;
                },
            });
        } catch (e) {
            // ignore
        }
    }

    return supportsPreventScrollCached;
}

function getScrollableElements(element: HTMLElement): ScrollableElement[] {
    const scrollableElements: ScrollableElement[] = [];
    const rootScrollingElement = document.scrollingElement || document.documentElement;

    let parent = element.parentNode;
    while (parent instanceof HTMLElement && parent !== rootScrollingElement) {
        if (parent.offsetHeight < parent.scrollHeight || parent.offsetWidth < parent.scrollWidth) {
            scrollableElements.push({
                element: parent,
                scrollTop: parent.scrollTop,
                scrollLeft: parent.scrollLeft,
            });
        }
        parent = parent.parentNode;
    }

    if (rootScrollingElement instanceof HTMLElement) {
        scrollableElements.push({
            element: rootScrollingElement,
            scrollTop: rootScrollingElement.scrollTop,
            scrollLeft: rootScrollingElement.scrollLeft,
        });
    }

    return scrollableElements;
}

function restoreScrollPosition(scrollableElements: ScrollableElement[]) {
    for (const { element, scrollTop, scrollLeft } of scrollableElements) {
        element.scrollTop = scrollTop;
        element.scrollLeft = scrollLeft;
    }
}

export function getActiveElement() {
    return document.activeElement as HTMLElement | null;
}

export function isElementInScope(scope: Element, element: Element) {
    return scope.contains(element);
}

export function isFocusFree() {
    return getActiveElement() === document.body;
}

export function createGuard(tabIndex: number) {
    const element = document.createElement('span');
    element.tabIndex = tabIndex;
    Object.assign(element.style, {
        width: '1px',
        height: 0,
        padding: 0,
        overflow: 'hidden',
        position: 'fixed',
        top: '1px',
        left: '1px',
    });
    element.setAttribute('aria-hidden', 'true');
    element.dataset[FOCUS_TRAP_GUARD_ATTR] = 'true';

    return element;
}

export function isGuard(node: HTMLElement) {
    return Boolean(node.dataset[FOCUS_TRAP_GUARD_ATTR]);
}

export function removeElementFromDocument(element: Element) {
    if (!element.parentNode) {
        return;
    }

    element.parentNode.removeChild(element);
}

function focusWithoutScrolling(element: HTMLElement) {
    if (supportsPreventScroll()) {
        element.focus({ preventScroll: true });
    } else {
        const scrollableElements = getScrollableElements(element);
        element.focus();
        restoreScrollPosition(scrollableElements);
    }
}

export function focusElement(element?: HTMLElement | null, preventScroll?: boolean) {
    if (!element) {
        return;
    }

    try {
        if (preventScroll) {
            focusWithoutScrolling(element);
        } else {
            element.focus();
        }
    } catch (e) {
        // ignore
    }
}

export function focusFirstIn(scope: HTMLElement) {
    const [first] = getTabbables(scope, (node) => !isGuard(node));

    focusElement(first);
}

export function getNextTabbableIndex(current: number, count: number, step: number) {
    if (current === -1) {
        return 0;
    }

    return count > 0 ? (count + current + step) % count : -1;
}
