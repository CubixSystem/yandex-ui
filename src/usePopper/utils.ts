import { Placement, BasePlacement, Offsets, top, left, right } from '@popperjs/core';

export function getLayoutRect(element: HTMLElement) {
    return {
        x: element.offsetLeft,
        y: element.offsetTop,
        width: element.offsetWidth,
        height: element.offsetHeight,
    };
}

export default function getBasePlacement(placement: Placement): BasePlacement {
    return placement.split('-')[0] as any;
}

export function distanceAndSkiddingToXY(placement: Placement, offset: [number, number]): Offsets {
    const basePlacement = getBasePlacement(placement);
    const invertDistance = [left, top].indexOf(basePlacement as any) >= 0 ? -1 : 1;

    const skidding = offset[0] || 0;
    const distance = (offset[1] || 0) * invertDistance;

    return [left, right].indexOf(basePlacement as any) >= 0
        ? { x: distance, y: skidding }
        : { x: skidding, y: distance };
}
