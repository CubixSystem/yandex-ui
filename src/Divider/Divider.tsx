import React, { FC, RefObject, CSSProperties, ElementType } from 'react';
import { cn } from '@bem-react/classname';

import { DividerLine, DividerLineProps } from './Line/Divider-Line';
import './Divider.css';

export interface DividerProps extends DividerLineProps {
    /**
     * Ссылка на корневой DOM-элемент компонента
     */
    innerRef?: RefObject<HTMLElement>;

    /**
     * Пользовательские стили
     */
    style?: CSSProperties;

    /**
     * Дополнительный класс
     */
    className?: string;

    /**
     * Компонент для рендера разделителя
     * @default 'div'
     */
    as?: ElementType;
}

export const cnDivider = cn('Divider');

/**
 * Компонент для отделения одной информации от другой.
 * @param {IDividerProps} props
 */
export const Divider: FC<DividerProps> = ({
    size = 1,
    color,
    innerRef,
    style,
    className,
    children,
    as: Component = 'div',
    ...props
}) => (
    <Component {...props} ref={innerRef} className={cnDivider(null, [className])} style={style}>
        <DividerLine size={size} color={color} />
        {children}
        {children && <DividerLine size={size} color={color} />}
    </Component>
);

Divider.displayName = cnDivider();
