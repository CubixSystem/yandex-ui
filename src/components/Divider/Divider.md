# Divider

<a href='https://github.yandex-team.ru/search-interfaces/frontend/tree/master/packages/lego-components/src/components/Divider' target='_blank'><img src='https://badger.yandex-team.ru/custom/[Исходники]/[Github
][green]/badge.svg' /></a> <a href='https://search.yandex-team.ru/stsearch?text=Divider.ts&facet.queue=ISL&facet.type=bug&facet.status=128' target='_blank'><img src='https://badger.yandex-team.ru/custom/[Известные проблемы]/[Startrek][blue]/badge.svg' /></a>

<!-- description:start -->
Компонент для отделения одной информации от другой.
<!-- description:end -->

## Пример использования

Конфигурация темы на уровне проекта:

```ts
// src/lib/theme.ts
import { configureRootTheme } from '@yandex-lego/components/Theme'
import { theme } from '@yandex-lego/components/Theme/presets/default'

configureRootTheme({ theme })
```

Использование компонента:

```ts
// src/App.ts
import React from 'react'
import { Divider } from '@yandex-lego/components/Divider'

const App = () => {
  return (
    <>
      Content
      <Divier style={{ margin: '8px 0' }} />
      Content
    </>
  )
}
```

## Примеры

### Простой разделитель

{{%story::desktop:content-divider-desktop--simple%}}

### Высота разделителя

Чтобы изменить размер разделителя, установите свойство `size` в необходимое значение.

{{%story::desktop:content-divider-desktop--size%}}

### Цвет разделителя

Чтобы изменить цвет разделителя, установите свойство `color` в необходимое значение.

{{%story::desktop:content-divider-desktop--color%}}

### Разделитель с текстом

{{%story::desktop:content-divider-desktop--with-text%}}

### Разделитель с пользовательскими стилями

Чтобы изменить внешние или внутренние отступы, установите свойство `style` с необходимыми параметрами.

{{%story::desktop:content-divider-desktop--with-styles%}}

## Свойства

<!-- props:start -->
| Свойство   | Тип                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                        | По умолчанию | Описание                                  |
| ---------- | -------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- | ------------ | ----------------------------------------- |
| innerRef?  | `RefObject<HTMLDivElement>`                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                | —            | Ссылка на корневой DOM-элемент компонента |
| size?      | `number`                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                   | `1`          | Высота разделителя в пикселях              |
| style?     | `CSSProperties`                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                            | —            | Пользовательские стили                    |
| className? | `string`                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                   | —            | Дополнительный класс                      |
| color?     | `string`                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                   | —            | Цвет разделителя                          |
| as?        | `"symbol" \| "object" \| "a" \| "abbr" \| "address" \| "area" \| "article" \| "aside" \| "audio" \| "b" \| "base" \| "bdi" \| "bdo" \| "big" \| "blockquote" \| "body" \| "br" \| "button" \| "canvas" \| "caption" \| "cite" \| "code" \| "col" \| "colgroup" \| "data" \| "datalist" \| "dd" \| "del" \| "details" \| "dfn" \| "dialog" \| "div" \| "dl" \| "dt" \| "em" \| "embed" \| "fieldset" \| "figcaption" \| "figure" \| "footer" \| "form" \| "h1" \| "h2" \| "h3" \| "h4" \| "h5" \| "h6" \| "head" \| "header" \| "hgroup" \| "hr" \| "html" \| "i" \| "iframe" \| "img" \| "input" \| "ins" \| "kbd" \| "keygen" \| "label" \| "legend" \| "li" \| "link" \| "main" \| "map" \| "mark" \| "menu" \| "menuitem" \| "meta" \| "meter" \| "nav" \| "noindex" \| "noscript" \| "ol" \| "optgroup" \| "option" \| "output" \| "p" \| "param" \| "picture" \| "pre" \| "progress" \| "q" \| "rp" \| "rt" \| "ruby" \| "s" \| "samp" \| "script" \| "section" \| "select" \| "small" \| "source" \| "span" \| "strong" \| "style" \| "sub" \| "summary" \| "sup" \| "table" \| "tbody" \| "td" \| "textarea" \| "tfoot" \| "th" \| "thead" \| "time" \| "title" \| "tr" \| "track" \| "u" \| "ul" \| "var" \| "video" \| "wbr" \| "webview" \| "svg" \| "animate" \| "animateMotion" \| "animateTransform" \| "circle" \| "clipPath" \| "defs" \| "desc" \| "ellipse" \| "feBlend" \| "feColorMatrix" \| "feComponentTransfer" \| "feComposite" \| "feConvolveMatrix" \| "feDiffuseLighting" \| "feDisplacementMap" \| "feDistantLight" \| "feDropShadow" \| "feFlood" \| "feFuncA" \| "feFuncB" \| "feFuncG" \| "feFuncR" \| "feGaussianBlur" \| "feImage" \| "feMerge" \| "feMergeNode" \| "feMorphology" \| "feOffset" \| "fePointLight" \| "feSpecularLighting" \| "feSpotLight" \| "feTile" \| "feTurbulence" \| "filter" \| "foreignObject" \| "g" \| "image" \| "line" \| "linearGradient" \| "marker" \| "mask" \| "metadata" \| "mpath" \| "path" \| "pattern" \| "polygon" \| "polyline" \| "radialGradient" \| "rect" \| "stop" \| "switch" \| "text" \| "textPath" \| "tspan" \| "use" \| "view" \| ComponentClass<any, any> \| FunctionComponent<any>` | `'div'`      | Компонент для рендера разделителя         |
<!-- props:end -->
