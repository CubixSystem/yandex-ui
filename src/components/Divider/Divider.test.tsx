import React, { createRef, ComponentType } from 'react';
import { mount, render } from 'enzyme';
import { ExtractProps } from '@bem-react/core';

import { Divider as ProgressCommon } from './Divider';
import { serverEnvironmentSetup, delay } from '../../internal/utils';

const platforms = [['desktop', ProgressCommon], ['touch-pad', ProgressCommon], ['touch-phone', ProgressCommon]];

type ProgressProps = ExtractProps<typeof ProgressCommon>;

describe.each<any>(platforms)('Divider@%s', (_platform, Divider: ComponentType<ProgressProps>) => {
    describe('server environment', () => {
        serverEnvironmentSetup();

        test('должен вернуть полный шаблон компонента (snapshot)', () => {
            expect(render(<Divider className="mix" />)).toMatchSnapshot();
        });
    });

    describe('client environment', () => {
        test('должен вернуть полный шаблон компонента (snapshot)', () => {
            expect(render(<Divider className="mix" />)).toMatchSnapshot();
        });

        test('должен быть установлен корректный displayName', () => {
            expect(Divider.displayName).toBe('Divider');
        });

        test('должен устанавливать ссылку на корневой DOM элемент', async () => {
            const innerRef = createRef<HTMLDivElement>();
            mount(<Divider innerRef={innerRef} />);
            await delay(100);
            expect(innerRef.current).not.toBe(null);
        });

        test('должен устанавливать и изменять size через props', () => {
            const wrapper = mount(<Divider size={2} />);
            expect(wrapper.prop('size')).toBe(2);
            wrapper.setProps({ size: 10 });
            expect(wrapper.prop('size')).toBe(10);
        });

        test('должен устанавливать и изменять color через props', () => {
            const wrapper = mount(<Divider />);
            wrapper.setProps({ color: '#fc0' });
            expect(wrapper.prop('color')).toBe('#fc0');
        });

        test('должен изменять тег компонента через props as', () => {
            const wrapper = mount(<Divider as="span" />);
            expect(wrapper.find('span')).toHaveLength(2);
        });
    });
});
