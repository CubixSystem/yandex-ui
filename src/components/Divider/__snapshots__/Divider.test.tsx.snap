// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Divider@desktop client environment должен вернуть полный шаблон компонента (snapshot) 1`] = `
<div
  class="Divider mix"
>
  <span
    class="Divider-Line"
    style="height:1px"
  />
</div>
`;

exports[`Divider@desktop server environment должен вернуть полный шаблон компонента (snapshot) 1`] = `
<div
  class="Divider mix"
>
  <span
    class="Divider-Line"
    style="height:1px"
  />
</div>
`;

exports[`Divider@touch-pad client environment должен вернуть полный шаблон компонента (snapshot) 1`] = `
<div
  class="Divider mix"
>
  <span
    class="Divider-Line"
    style="height:1px"
  />
</div>
`;

exports[`Divider@touch-pad server environment должен вернуть полный шаблон компонента (snapshot) 1`] = `
<div
  class="Divider mix"
>
  <span
    class="Divider-Line"
    style="height:1px"
  />
</div>
`;

exports[`Divider@touch-phone client environment должен вернуть полный шаблон компонента (snapshot) 1`] = `
<div
  class="Divider mix"
>
  <span
    class="Divider-Line"
    style="height:1px"
  />
</div>
`;

exports[`Divider@touch-phone server environment должен вернуть полный шаблон компонента (snapshot) 1`] = `
<div
  class="Divider mix"
>
  <span
    class="Divider-Line"
    style="height:1px"
  />
</div>
`;
